portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFolderLocalServiceImpl.java

```diff
@@ -1425,6 +1438,86 @@ protected void validateFolderName(String folderName)
 		}
 	}
 
+	private boolean _hasInheritablePermissions(DLFolder dlFolder) {
+		long parentFolderId = dlFolder.getParentFolderId();
+
+		if (parentFolderId == DLFolderConstants.DEFAULT_PARENT_FOLDER_ID) {
+			int count =
+				_resourcePermissionLocalService.getResourcePermissionsCount(
+					dlFolder.getCompanyId(), DLConstants.SERVICE_NAME,
+					ResourceConstants.SCOPE_INDIVIDUAL,
+					String.valueOf(dlFolder.getGroupId()));
+
+			if (count == 0) {
+				return false;
+			}
+
+			parentFolderId = dlFolder.getGroupId();
+		}
+
+		// TODO Update the check permission in here
+
+		PermissionPropagationEntry permissionPropagationEntry =
+			_permissionPropagationEntryLocalService.
+				fetchPermissionPropagationEntry(
+					dlFolder.getCompanyId(), dlFolder.getGroupId(),
+					DLFolderConstants.getClassName(), parentFolderId);
+
+		if (permissionPropagationEntry == null) {
+			return false;
+		}
+
+		return permissionPropagationEntry.isPropagation();
+	}
+
+	private void _inheritRolesPermissions(DLFolder dlFolder)
+		throws PortalException {
+
+		long companyId = dlFolder.getCompanyId();
+
+		long parentFolderId = dlFolder.getParentFolderId();
+
+		String parentClassName = DLFolderConstants.getClassName();
+
+		if (parentFolderId == DLFolderConstants.DEFAULT_PARENT_FOLDER_ID) {
+			parentClassName = DLConstants.RESOURCE_NAME;
+			parentFolderId = dlFolder.getGroupId();
+		}
+
+		Set<String> dlFolderActionIds = SetUtil.fromCollection(
+			ResourceActionsUtil.getModelResourceActions(
+				DLFolderConstants.getClassName()));
+
+		dlFolderActionIds.add(ActionKeys.ADD_FOLDER);
```

Comment:
I think ActionKeys.ADD_FOLDER should only be added when the parent folder is Home. Could refactor like:
```
		long companyId = dlFolder.getCompanyId();

		String parentClassName = DLFolderConstants.getClassName();

		Set<String> dlFolderActionIds = SetUtil.fromCollection(
			ResourceActionsUtil.getModelResourceActions(parentClassName));

		long parentFolderId = dlFolder.getParentFolderId();

		if (parentFolderId == DLFolderConstants.DEFAULT_PARENT_FOLDER_ID) {
			parentClassName = DLConstants.RESOURCE_NAME;
			parentFolderId = dlFolder.getGroupId();
			dlFolderActionIds.add(ActionKeys.ADD_FOLDER);
		}
```